pkg/controller/machine.go

```diff
@@ -266,14 +266,39 @@ func (c *controller) getMachineFromNode(nodeName string) (*v1alpha1.Machine, err
 }
 
 func (c *controller) updateMachineState(machine *v1alpha1.Machine) (*v1alpha1.Machine, error) {
+	nodeName := machine.Status.Node
 
-	if machine.Status.Node == "" {
-		// There are no objects mapped to this machine object
-		// Hence node status need not be propogated to machine object
-		return machine, nil
+	if nodeName == "" {
+		// Check if any existing node-object can be adopted.
+		nodeList, err := c.nodeLister.List(labels.Everything())
+		if err != nil {
+			klog.Errorf("Could not list the nodes due to error: %v", err)
+			return machine, err
+		}
+		for _, node := range nodeList {
+			nID, mID := decodeMachineID(node.Spec.ProviderID), decodeMachineID(machine.Spec.ProviderID)
+			if nID == mID {
+				klog.V(2).Infof("Adopting the node object %s for machine %s", node.Name, machine.Name)
+				nodeName = node.Name
+				clone := machine.DeepCopy()
+				clone.Status.Node = nodeName
+				clone, err = c.controlMachineClient.Machines(clone.Namespace).UpdateStatus(clone)
```

Comment:
>We need to update clone.Labels["node"], .status.LastOperation and .status.CurrentStatus also here.

`clone.Labels["node"]` will be updated [here](https://github.com/gardener/machine-controller-manager/blob/859ef7504d2ff3efa715bfd6c84317c3267de835/pkg/controller/machine.go#L353) in the same function. 

I couldn't understand why `LastOperation` or `CurrentStatus` needs to be updated here, can you please elaborate. Basically, normal machine-controller flow should apply, this part of the code should execute only when `status.Node` is missing for any reason- possibly when MCM restarts during machineCreate() and updating the node-name to machine-object failed. As at the moment, our only source of node-name is from the `driver.Create()`, the output of which might be lost, we decided to adopt the existing node-object via providerID.

Reply:
>In case this change is also supposed to take care of https://github.com/gardener/machine-controller-manager/issues/423#issuecomment-597628323, then mID will not be available as mentioned in #421

I couldn't understand this, can you please explain a little more. Basically the specific changes here of adopting-node are for the creation-flow.

Reply:
May be i need help with understanding the scenario.
Consider the following case:
T1: machine is created
T2. MCM processes fresh machine-objects by making a cloud call of CreateVMs().
- Machine.Status and Spec.ProviderID is set only after this Create call returns.

T2. Before CreateVM() call returns, MCM is brought down/crashed
T3. Node joins
T4. MCM comes up after several retries (misses the node add event)
T5. `updateMachineState` is called by machine reconcile
- Does't see provider ID in spec

T6. `machineCreate` is called, and since machine exists, the provider ID is updated, and machine phase is set to "Pending"
T7. Machine is enqueued after 10 min
T7. Machine health timeout of 10 min (or less) expires and the machine is marked as failed

Is this a scenario that could happen?

Reply:
This is certainly a very good point and probably deserves more testing/experiments, as I  am not sure of the exact outcome of the above sequence. Would you be fine if we try to separately address this, outside from current PR?

Reply:
As discussed over call, we will address these concerns separately.